### PR TITLE
[Core] Add file name information collector

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This adds file name information collector, which is capable of generating file names based on pattern string, or finding and sorting files based on the patterns. Sorting can be given priority rank, step or time_step value.

```c++
     * Construct an object which will handle given filename pattern. This rFileNamePattern
     * can incldue folders, subfolders. It must include the file name. Followings are
     * the list of flags supported.
     *
     * Accepted flags are:
     *      "<rank>"                : ranks of the model part
     *      "<step>"                : STEP value in model parts's process info
     *      "<time>"                : TIME value in model part's process info
     *      "<model_part_name>"     : name of the model part
     *      "<model_part_full_name>": full name of the model part.
     *
     * Flags "<rank>", "<step>" and "<time>" is only allowed be in the file name only. (Those flags are not
     * allowed to be present in file path)
     *
     * Example rFileNamePattern:
     *      1. "test_cases/<model_part_name>/<model_part_full_name>-<time>.h5"
     *      2. "test/test_1_<rank>_<time>.h5"
     *
     * Example rTimeStepFormat:
     *      1. "0.4f"
     *      2. "%0.4f"
     *      3. "0.3e"
```

Following patters are tested:
```c++
    "/test_folder/<model_part_full_name>/<model_part_name>-<time>.h5"
    "/test_folder/<model_part_full_name>/<time>-<model_part_name>.h5"
    "/test_folder/<model_part_full_name>/<time>-<rank>-<model_part_name>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>-<rank>-<step>-<time>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>-<rank>-<time>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>_<time>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>_<rank>_<step>_<time>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>_<rank>_<time>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>_<rank>_<step>-<time>.h5"
    "/test_folder/<model_part_full_name>/<model_part_name>_<rank>-<time>.h5"
    "/test_folder/<model_part_full_name>/test_model_part_stat_<rank>_<time>.h5"
```

**Changelog**
- Add `FileNameInformationCollector`
- Exposing to python
- Add tests

### CODE REVIEW

File: kratos/includes/kratos_filesystem.h [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 // Project includes
 #include "includes/define.h"
-
+#include "includes/model_part.h"
```
Comment (loumalouomega) [inline kratos/includes/kratos_filesystem.h]: Maybe you can do a forward declaration to avoid the include, opinion @philbucher ?

File: kratos/includes/kratos_filesystem.h [hunk 2]
```diff
@@ -48,6 +48,185 @@ std::uintmax_t KRATOS_API(KRATOS_CORE) remove_all(const std::string& rPath);
...
+        std::vector<FileNameData>& rFileNameDataList,
+        const std::vector<std::string>& rFlagsSortingOrder);
+
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    enum DataType {
```
Comment (loumalouomega) [inline kratos/includes/kratos_filesystem.h]: enum class please

File: kratos/tests/cpp_tests/sources/test_filesystem.cpp [hunk 3]
```diff
@@ -15,8 +15,11 @@
 #include <fstream>
 
 // External includes
+#include "ghc/filesystem.hpp" // TODO after moving to C++17 this can be removed since the functions can be used directly
```
Comment (loumalouomega) [inline kratos/tests/cpp_tests/sources/test_filesystem.cpp]: In theory this is included in #include "includes/kratos_filesystem.h"

Comment (loumalouomega) [inline kratos/tests/cpp_tests/sources/test_filesystem.cpp]:
```suggestion
```

**Review by loumalouomega — COMMENTED** [review]
Some comments, good idea

**Comment by loumalouomega** [comment]
Our arch enemy windows CI attacks again

File: kratos/tests/cpp_tests/sources/test_filesystem.cpp (hunk 3 above)
Comment (sunethwarna) [reply kratos/tests/cpp_tests/sources/test_filesystem.cpp]: removed :)

Comment (sunethwarna) [reply kratos/tests/cpp_tests/sources/test_filesystem.cpp]: removed :)

File: kratos/includes/kratos_filesystem.h (hunk 2 above)
Comment (sunethwarna) [reply kratos/includes/kratos_filesystem.h]: removed :)

**Review by loumalouomega — COMMENTED** [review]
OK from my side

File: external_libraries/ghc/filesystem.hpp [hunk 4]
```diff
@@ -102,6 +102,7 @@
 #ifdef GHC_EXPAND_IMPL
 
 #ifdef GHC_OS_WINDOWS
+#define NOMINMAX
```
Comment (philbucher) [inline external_libraries/ghc/filesystem.hpp]: can you please revert the changes in this file? Should no longer be necessary

Comment (sunethwarna) [reply external_libraries/ghc/filesystem.hpp]: Are you sure? Windows build was failing because of not having this line...

Comment (philbucher) [reply external_libraries/ghc/filesystem.hpp]: you need to exclude the cpp that includes this from Cotire

in any case including this header in another header is nogo bcs it includes `windows.h`

Comment (sunethwarna) [reply external_libraries/ghc/filesystem.hpp]: Hmm... I am not including this header in another header, but I am including this header in another cpp.... I'll try removing changes.

Comment (sunethwarna) [reply external_libraries/ghc/filesystem.hpp]: done :)

**Comment by sunethwarna** [comment]
@philbucher this is ready for review :)

**Comment by philbucher** [comment]
@sunethwarna did you make sure this doesn't hang indefinetly in Win?
(We are really trying to have all tests working for the soon coming release)

**Comment by sunethwarna** [comment]
@philbucher Doing ryt now :)

**Comment by sunethwarna** [comment]
@philbucher Apparently now it works, except for the MMG crash. Could you take a look when you have time?

**Comment by sunethwarna** [comment]
@philbucher  I re-ran the CI to just to make sure, now the tests are passing in Windows CI as well (everything is integrated to small tests). I will start a CI re-run again just to be sure. :)

**Review by philbucher — DISMISSED** [review]
Thanks for the work @sunethwarna 
I checked it out locally, overall it looks good to me, but I will have to use it to fully understand how it works.
I think you will use it for #7618, I will check that out and then see

**Review by philbucher — APPROVED** [review]
